### DISCUSSION TITLE
libvips C++ api + glib errors

### DISCUSSION THREAD

**Discussion by bkocev** [libvips/libvips#3819] [General]
Hi,

the following test crashes on the line creating the image:

```C++
TEST_F( ImageWriterTest, testImageLoader ) {
    if( VIPS_INIT( "my_program" ) )
        vips_error_exit( NULL );

    auto image = vips::VImage::black( 100, 100 );

    vips_shutdown();
}
```

with the following error: 

```
Message: 

(proces.789: g_datalist_id_set_data_full: assertion 'key_id > 0' failed

(process:34332): GLib-GObject-CRITICAL **: 17:04:14.790: g_param_spec_pool_lookup: assertion 'pool != NULL' failed

(process:34332): GLib-GObject-CRITICAL **: 17:04:14.790: g_object_set_is_valid_property: object class '(NULL)' has no property named 'width'

(proces.790: g_datalist_id_set_data_full: assertion 'key_id > 0' failed

(process:34332): GLib-GObject-CRITICAL **: 17:04:14.790: g_param_spec_pool_lookup: assertion 'pool != NULL' failed

(process:34332): GLib-GObject-CRITICAL **: 17:04:14.790: g_object_set_is_valid_property: object class '(NULL)' has no property named 'height'

(proces.790: g_datalist_id_set_data_full: assertion 'key_id > 0' failed
unknown file: error: C++ exception with description "black: parameter width not set
black: parameter height not set
" thrown in the test body.

  Stack Trace: 
04:14 line 14
```

Any idea what the problem could be?

I have:

```cmake
conan_cmake_configure( REQUIRES glib/2.78.3
                                gtest/1.12.1                                
                                libvips/8.15.1                               
                       GENERATORS cmake_find_package
                       IMPORTS "bin, *.dll -> ${PROJECT_BINARY_DIR}/bin/"
                       OPTIONS "libvips:shared=True" "glib:shared=True")
```

and in my CMakelists:

```cmake
find_package(libvips REQUIRED)
```

and then linking to libvips::vips-cpp.

I have trouble detecting the real problem.

Thank you,
Bojan.

**Comment by jcupitt** [comment]
Hi @bkocev,

You can't mix compilers in a C++ project, so if you want to use the libvips C++ DLLs we supply, you must build with exactly the same compiler we used (gcc 10, from memory) and exactly the same flags. You can use `pkg-config` to get the flag set.

It's usually simpler to rebuild the C++ binding with your own C++ compiler. Just add this directory to your project:

https://github.com/libvips/libvips/tree/master/cplusplus

If you use meson it's very easy. With cmake you'll need to write your own build file.

**Reply by bkocev** [reply]
Hi @jcupitt ,

thank you for the reply. 
I am using Windows and Visual Studio 2022. 
As far as I could see on https://conan.io/center/recipes/libvips?version=8.15.1, Linux, Windows, macOS, and macOS Apple Silicon are all supported.
Interestingly, the C api works without problems. I only have issues with the C++ wrapper/api.
So, maybe there is still something that can be done?

Thank you,
Bojan.

**Reply by jcupitt** [reply]
That won't work -- C++ does not define an ABI, so you can't mix compilers. You'll need to rebuild the libvips C++ API using your VS version.
